Bignum: extract core_add from the prototype

tom-cosgrove-arm opened the pull request:
## Description
Extract MPI_CORE(add) from the [https://github.com/hanno-arm/mbedtls/blob/ecp_prototype/library/bignum_core.c#L181-L195](prototype) as per #6223 

## Status
**READY**

## Requires Backporting
**NO**

## Todos
- [x] The `MPI_CORE(add)` and its dependencies are taken from the prototype and credit is given to Hanno
- [x] New macros are expanded/removed
- [x] The legacy function in `bignum.c` (`mbedtls_mpi_add_abs()`) is calling the new function wherever possible
- [x] Extensive unit tests are added for all new functions

tom-cosgrove-arm commented:
It's just the last 3 commits - the earlier commits are from #6083, on which this builds

tom-cosgrove-arm commented on library/bignum.c:
```diff
@@ -934,68 +932,32 @@ int mbedtls_mpi_add_abs( mbedtls_mpi *X, const mbedtls_mpi *A, const mbedtls_mpi
-        tmp= *o;
-        *p +=  c; c  = ( *p <  c );
-        *p += tmp; c += ( *p < tmp );
-    }
```
upon leaving this loop, `i == j`

tom-cosgrove-arm commented on library/bignum.c:
```diff
@@ -934,68 +932,32 @@ int mbedtls_mpi_add_abs( mbedtls_mpi *X, const mbedtls_mpi *A, const mbedtls_mpi
     while( c != 0 )
     {
-        if( i >= X->n )
+        if( j >= X->n )
```
so switching from `i` to `j` here, to only use the one variable

tom-cosgrove-arm commented:
Rebased on top of `montmul`, which has had review comments applied. Still the actual work in this PR is only the last 3 commits.

tom-cosgrove-arm commented:
Rebased on top of the montmul PR #6083, which has had a doc comment fix applied. Still the actual work in this PR is only the last 3 commits.

tom-cosgrove-arm commented:
Rebased on top of the [montmul PR 6083](https://github.com/Mbed-TLS/mbedtls/pull/6083), which has had a doc comment fix applied. Still the actual work in this PR is only the last 3 commits.

tom-cosgrove-arm commented:
Rebased on top of the latest montmul. Still the actual work in this PR is only the last 3 commits.

gilles-peskine-arm commented:
Now that https://github.com/Mbed-TLS/mbedtls/pull/6373 is merged, I think this can be rebased on top of development.

tom-cosgrove-arm commented:
Rebased on top of [PR 6399 - Split bignum tests by module and generate with Python framework](https://github.com/Mbed-TLS/mbedtls/pull/6399)

tom-cosgrove-arm commented on tests/suites/test_suite_bignum_core.function:
```diff
@@ -339,67 +339,101 @@ exit:
     TEST_EQUAL( 0, mbedtls_test_read_mpi_core( &B, &B_limbs, input_B ) );
     TEST_EQUAL( 0, mbedtls_test_read_mpi_core( &S, &S_limbs, input_S ) );
-    X_limbs = S_limbs;
-    ASSERT_ALLOC( X, X_limbs );
```
move this down, to after we've checked that the limb sizes are all the same, and dispense with `X_limbs`

wernerlewis reviewed (commented):
One minor comment on consistency in documentation, otherwise changes look good to me.

minosgalanakis commented on library/bignum.c:
```diff
@@ -925,27 +924,25 @@ int mbedtls_mpi_add_abs( mbedtls_mpi *X, const mbedtls_mpi *A, const mbedtls_mpi
-        *p += tmp; c += ( *p < tmp );
-    }
+    mbedtls_mpi_uint *p = X->p;
+
```
Not blocker, but wondering on why are we adding newline between those stetements (lines 930, 932)

tom-cosgrove-arm replied:
I tried it with and without blank lines in various places, and to my eye this arrangement looked best.

minosgalanakis commented on library/bignum.c:
```diff
@@ -925,27 +924,25 @@ int mbedtls_mpi_add_abs( mbedtls_mpi *X, const mbedtls_mpi *A, const mbedtls_mpi
-    }
+    mbedtls_mpi_uint *p = X->p;
+
+    mbedtls_mpi_uint c = mbedtls_mpi_core_add( p, p, B->p, j);
```
Whitespace

`mbedtls_mpi_core_add( p, p, B->p, j)`  -> `mbedtls_mpi_core_add( p, p, B->p, j )`

minosgalanakis reviewed (commented):
Looks good overall, minor comments on style.

tom-cosgrove-arm commented:
Rebased on top of latest `development`, and addressed feedback given privately by Janos:

> I think it is Ok to reuse the test cases for add, but would be helpful for the reader to separate the test function. I mean the main test function would do the setup for the parameters and would call a function that does the `add` tests and then call another that performs the `add_if` tests.

wernerlewis reviewed (approved):
Changes LGTM. I have one comment on return values for the add/add_if test functions, but this isn't a blocker.

gilles-peskine-arm commented on tests/suites/test_suite_bignum_core.function:
```diff
@@ -5,6 +5,154 @@
+ * \param carry       Expected carry from the addition.
+ * \param[in,out] X   Temporary storage to be used for results.
+ *
+ * \return  1 if mbedtls_mpi_core_add() passes this test, otherwise 0.
```
My bad. We do have a habit of returning pass/fail booleans rather than status codes from auxiliary functions. I'm afraid I started this trend in PSA tests, where a status code was `psa_status_t` and thus it was natural to give the type `int` its habitual C meaning of true/false indicating ok/not-ok. Then we've carried this over to non-PSA tests where `int` is ambiguous. And now we have a mix of conventions.

tom-cosgrove-arm replied:
In the library this would be a blocker - mixed use of "0 means success" and "truthy (non-zero) means success" have been famously confusing in OpenSSL - but in test code we can live with it.

minosgalanakis reviewed (approved):
LGTM
